src/Compiler/Facilities/AsyncMemoize.fs

```diff
@@ -153,409 +141,108 @@ type internal AsyncMemoize<'TKey, 'TVersion, 'TValue when 'TKey: equality and 'T
-                                let currentLogger = DiagnosticsThreadStatics.DiagnosticsLogger
-                                DiagnosticsThreadStatics.DiagnosticsLogger <- cachingLogger
-
-                                log (Started, key)
-
-                                try
-                                    let! result = computation
-                                    post (key, (JobCompleted(result, cachingLogger.CapturedDiagnostics)))
-                                    return result
-                                finally
-                                    DiagnosticsThreadStatics.DiagnosticsLogger <- currentLogger
-                            },
-                            cancellationToken = linkedCtSource.Token
-                        )
-                        |> Async.AwaitTask
-                with
-                | TaskCancelled ex ->
-                    // TODO: do we need to do anything else here? Presumably it should be done by the registration on
-                    // the cancellation token or before we triggered our own cancellation
+            otherVersions,
 
-                    // Let's send this again just in case. It seems sometimes it's not triggered from the registration?
+            cached
+            |> Option.map countHit
+            |> Option.defaultWith cacheSetNewJob
 
-                    Interlocked.Increment &cancel_exception_original |> ignore
+        async {            
+            let otherVersions, job = lock cache getOrAdd
 
-                    post (key, (OriginatorCanceled))
-                    return raise ex
-                | ex ->
-                    post (key, (JobFailed(ex, cachingLogger.CapturedDiagnostics)))
-                    return raise ex
+            log Requested key
+
+            if cancelDuplicateRunningJobs && not cancelUnawaitedJobs then
+                otherVersions |> Seq.map snd |> Seq.iter _.CancelIfUnawaited()
 
-            | Existing job -> return! job |> Async.AwaitTask
+            use _ = new CompilationGlobalsScope()
 
+            let! result, logger = job.Request
+            logger.CommitDelayedDiagnostics DiagnosticsThreadStatics.DiagnosticsLogger
+            match result with
+            | Ok result ->
+                return result
+            | Error ex ->
+                return raise ex
         }
 
     member _.TryGet(key: 'TKey, predicate: 'TVersion -> bool) : 'TValue option =
-        let versionsAndJobs = cache.GetAll(key)
+        let versionsAndJobs =
+            lock cache <| fun () ->cache.GetAll(key)
 
         versionsAndJobs
         |> Seq.tryPick (fun (version, job) ->
-            match predicate version, job with
-            | true, Completed(completed, _) -> Some completed
+            match predicate version, job.Result with
+            | true, Some(Ok result, _) -> Some result
             | _ -> None)
 
     member _.Clear() = cache.Clear()
```

Comment:
It wasn't there before, but probably this should be guarded by a lock? Especially the one with `predicate`.

Reply:
Yes, and job.Result should be renamed TryResult here because it suggests Task.Result but in reality it is an option.